crates/rspack_core/src/build_chunk_graph/incremental.rs

```diff
@@ -0,0 +1,605 @@
+    if !cgi.min_available_modules_init
+      || !self.hit_cache(
+        cache,
+        &cgi.runtime,
+        cgi.min_available_modules.clone(),
+        block.get_group_options(),
+      )
+    {
+      self.stat_cache_miss_by_available_modules += 1;
+      return false;
+    }
+
+    let cache_result = cache
+      .cache_result
+      .as_ref()
+      .expect("should have cache result");
+
+    // update cache available modules
+    self.outdated_chunk_group_info.insert(cgi_ukey);
+
+    let cgi = self.chunk_group_infos.expect_get_mut(&cgi_ukey);
+    cgi.skipped_items = cache_result.skipped_modules.clone();
+
+    let chunk_graph = &mut compilation.chunk_graph;
+    for module in &cache_result.modules {
+      let ordinal = self.get_module_ordinal(*module);
+      chunk_graph.connect_chunk_and_module(chunk, *module);
+      let mask = self.mask_by_chunk.entry(chunk).or_default();
+      mask.set_bit(ordinal, true);
+
+      // TODO: correct preorder index
+    }
+
+    for block in &cache_result.outgoings {
+      self.make_chunk_group(
+        *block,
+        *self.edges.get(block).expect("should have module for block"),
+        cgi_ukey,
+        chunk,
+        compilation,
+      );
+    }
+
+    true
+  }
+
+  #[instrument(skip_all)]
+  pub fn update_with_compilation(&mut self, compilation: &mut Compilation) -> Result<()> {
+    let enable_incremental = compilation
+      .incremental
+      .can_read_mutations(IncrementalPasses::BUILD_CHUNK_GRAPH);
+
+    // This optimization is from  https://github.com/webpack/webpack/pull/18090 by https://github.com/dmichon-msft
+    // Thanks!
+    let module_graph = compilation.get_module_graph();
+    let ordinal_by_module = &mut self.ordinal_by_module;
+    for m in module_graph.modules().keys() {
+      if !ordinal_by_module.contains_key(m) {
+        ordinal_by_module.insert(*m, ordinal_by_module.len() as u64 + 1);
+      }
+    }
+    for chunk in compilation.chunk_by_ukey.keys() {
+      let mut mask = BigUint::from(0u32);
+      for module in compilation
+        .chunk_graph
+        .get_chunk_modules(chunk, &module_graph)
+      {
+        let module_id = module.identifier();
+        let module_ordinal = self.get_module_ordinal(module_id);
+        mask.set_bit(module_ordinal, true);
+      }
+      self.mask_by_chunk.insert(*chunk, mask);
+    }
+
+    if !enable_incremental {
+      return Ok(());
+    }
+
+    self.stat_invalidated_chunk_group = 0;
+    self.stat_invalidated_caches = 0;
+    self.stat_use_cache = 0;
+    self.stat_chunk_group_created = 0;
+    self.stat_cache_miss_by_available_modules = 0;
+    self.stat_cache_miss_by_cant_rebuild = 0;
+
+    let modules = if compilation
+      .incremental
+      .can_read_mutations(IncrementalPasses::MAKE)
+    {
+      let affected_lock = compilation
+        .incremental
+        .mutations_read(IncrementalPasses::MAKE)
+        .expect("should get mutation")
+        .get_affected_modules_with_module_graph(&module_graph);
```

Comment:
```suggestion
    let modules = if let Some(mutations) = compilation
      .incremental
      .mutations_write()
    {
      let affected_lock = mutations.get_affected_modules_with_module_graph(&module_graph);
```

Reply:
And if you have the write permission, then you need to collect all the mutations in your pass, currently we don't need to collect these for chunk graph, we have a separate global effect pass to find the changed modules by compare the all modules' blocks and module_id, but we might need to collect them if we can remove the separate pass